### PR TITLE
[FIX] account: add default country_id on account_tax

### PR DESCRIPTION
**Pull request by kdes-odoo**
### Before this commit
---

During chart template loading, `account.tax.group` records are created with an explicit `country_id` defined in [data file](https://github.com/odoo/odoo/blob/be685237124b4c645989811eac9b9ae115d1c8d9/addons/l10n_sa/data/template/account.tax.group-sa.csv#L1), ensuring each tax group is linked to the localization's country.

However, `account.tax` records are created without an explicit `country_id`. This triggers the `country_id` [compute](https://github.com/odoo/odoo/blob/5de4a5f6caee09881e7e147062f9eab9f8916a61/addons/account/models/account_tax.py#L222-L224), which first assigns `account_fiscal_country_id` to the tax.

For example, if the **Saudi Arabia** localization is loaded while the fiscal country is set to **India**, tax groups are created with the Saudi Arabia country, but taxes are assigned the Indian country through the compute.

image

If any [migration script like in l10n_sa](https://github.com/odoo/odoo/blob/4cb3090ee829084c2a9260b6d38095cd818678f5/addons/l10n_sa/migrations/2.1/end-migrate.py#L34-L36) tries to load the chart template, As a result, the tax and its tax group end up with different `country_id` values, causing the [validation](https://github.com/odoo/odoo/blob/5de4a5f6caee09881e7e147062f9eab9f8916a61/addons/account/models/account_tax.py#L215-L219) that requires both records to belong to the same country to fail.

This will happen for all localization's chart_template load and is not an localization specific.

### Traceback

```python
File "/home/odoo/src/odoo/19.0/addons/l10n_sa/migrations/2.1/end-migrate.py", line 36, in migrate
    env["account.chart.template"].try_loading("sa", company)
  File "/home/odoo/src/odoo/19.0/addons/account/models/chart_template.py", line 171, in try_loading
    return self._load(template_code, company, install_demo, force_create)
  File "/home/odoo/src/odoo/19.0/addons/account/models/chart_template.py", line 237, in _load
    self._load_data(data)
  File "/tmp/tmpl8iuyhtm/migrations/account/0.0.0/pre-ensure-deferred-accounts.py", line 37, in _load_data
    return super()._load_data(data, *args, **kwargs)
  File "/home/odoo/src/odoo/19.0/addons/account/models/chart_template.py", line 697, in _load_data
    created_records[model] = self.with_context(lang='en_US').env[model]._load_records(all_records_vals)
  File "/home/odoo/src/odoo/19.0/odoo/orm/models.py", line 5196, in _load_records
    records = self._load_records_create([data['values'] for data in to_create])
  File "/home/odoo/src/odoo/19.0/odoo/orm/models.py", line 5103, in _load_records_create
    records = self.create(vals_list)
  File "/home/odoo/src/odoo/19.0/odoo/orm/decorators.py", line 369, in create
    return method(self, vals_list)
  File "/home/odoo/src/odoo/19.0/addons/account/models/account_tax.py", line 668, in create
    taxes = super(AccountTax, self.with_context(context)).create([self._sanitize_vals(vals) for vals in vals_list])
  File "/home/odoo/src/odoo/19.0/odoo/orm/decorators.py", line 369, in create
    return method(self, vals_list)
  File "/home/odoo/src/odoo/19.0/addons/mail/models/mail_thread.py", line 324, in create
    threads = super(MailThread, self).create(vals_list)
  File "/home/odoo/src/odoo/19.0/odoo/orm/decorators.py", line 369, in create
    return method(self, vals_list)
  File "/tmp/tmpl8iuyhtm/migrations/util/orm.py", line 267, in wrapper
    return f(*args, **kwargs)
  File "/tmp/tmpl8iuyhtm/migrations/base/0.0.0/pre-models-match_uniq.py", line 113, in create
    records = super().create([vals_list[idx] for idx in create_idx_list])
  File "/home/odoo/src/odoo/19.0/odoo/orm/decorators.py", line 369, in create
    return method(self, vals_list)
  File "/home/odoo/src/odoo/19.0/odoo/orm/models.py", line 4711, in create
    records = self._create(data_list)
  File "/home/odoo/src/enterprise/19.0/ai_fields/models/models.py", line 43, in _create
    return super()._create(data_list)
  File "/home/odoo/src/odoo/19.0/odoo/orm/models.py", line 4949, in _create
    records._validate_fields(name for data in data_list for name in data['stored'])
  File "/home/odoo/src/odoo/19.0/odoo/orm/models.py", line 1268, in _validate_fields
    check(records)
  File "/home/odoo/src/odoo/19.0/addons/account/models/account_tax.py", line 245, in validate_tax_group_id
    raise ValidationError(_("The tax group must have the same country_id as the tax using it."))
odoo.exceptions.ValidationError: The tax group must have the same country_id as the tax using it.
```

### After this commit
---

When creating an `account.tax` record, we first check whether `country_id` is explicitly provided.

If it is not provided and `tax_group_id` is present, we assign the tax group's `country_id` to the creation values instead of triggering the compute.

This ensures taxes inherit the same country as their tax group and prevents validation errors during chart template loading.

OPW: 6313209
Upgrade Traceback group : 2096

### CODE REVIEW

File: addons/account/models/account_tax.py [hunk 1]
```diff
@@ -535,6 +535,21 @@ def _sanitize_vals(self, vals):
...
+            vals['tax_group_id']
+            for vals in vals_list
+            if not vals.get('country_id') and vals.get('tax_group_id')
+        }
+        if tax_group_ids:
+            tax_group_countries = {
+                tg.id: tg.country_id.id
+                for tg in self.env['account.tax.group'].browse(tax_group_ids)
+                if tg.country_id
+            }
```
Comment (epictete) [inline addons/account/models/account_tax.py]:
```suggestion
            grouped_tax_groups = self.env['account.tax.group'].browse(tax_group_ids).filtered('country_id').grouped('id')
```
Nit, purely cosmetic (feel free to ignore) ☺️

File: addons/account/models/account_tax.py [hunk 2]
```diff
@@ -535,6 +535,21 @@ def _sanitize_vals(self, vals):
 
     @api.model_create_multi
     def create(self, vals_list):
+        tax_group_ids = {
```
Comment (epictete) [inline addons/account/models/account_tax.py]: I think it would be nice to add a comment here, explaining why this needs to be done.
It's not obvious and could be helpful to future developers (maybe you 😉) dealing with this code block 🤓

File: addons/account/models/account_tax.py (hunk 1 above)
Comment (kdes-odoo) [reply addons/account/models/account_tax.py]: Lets just work with ids rather then storing the recordset. keeping it simpler. :wink:

**Comment by kdes-odoo** [comment]
Hey @epictete !!

Thanks for your initial review, I've added the comment, Could you please review the changes and let me know if there's anything else that needs to be updated on my side?

**Comment by kdes-odoo** [comment]
> Thanks for the quick update 👌 But I was thinking that it would probably make more sense to simply update `_compute_country_id()`, so that the tax group country takes precedence, wdyt?

@epictete , yes, I agree with you. 🙂 My hesitation is that, in accounting, things are usually implemented for a reason, so I prefer avoiding structural changes like updating the compute method unless they're really necessary. Also, since it's part of the depends on [`company_id`](https://github.com/odoo/odoo/blame/f59783d43f891fea14d03f67d7c017d7ffbbe5a7/addons/account/models/account_tax.py#L278), I think it's better not to change the precedence.

With the current approach, we're addressing the specific issue we're facing without altering the existing computation logic.

That said, if you feel updating `_compute_country_id()` is the better long-term solution here, I'm happy to update the fix. I'd like to know your perspective.

**Comment by epictete** [comment]
Hey @kdes-odoo 
I think we should update the compute method.
It does not make sense in any case to have a different country on a tax / tax group, which is why it is enforced by a constraint.
So let's align `_compute_country_id()` with the constraint :ok_hand:

**Comment by kdes-odoo** [comment]
Hey @epictete !!! 

Thanks for your confirmation on updating the compute function itself, I've updated the compute and tested the CoA load and also migrations edge case where it was failing, those cases are successfully managed, but now we have few test cases failing, Here is overview of my analysis : 

There might be the tax that does not has associated tax group to it, I.e. https://github.com/odoo/odoo/blob/19039d174348660007f6d8f23311620f94acef68/addons/l10n_cz/data/template/account.tax-cz.csv#L142

For this, with new compute this will be the call stack : 
```python
  /home/odoo/odoo19/community/19.0/addons/account/models/account_tax.py(281)_compute_country_id()
-> tax.country_id = tax.tax_group_id.country_id or tax.company_id.account_fiscal_country_id or tax.company_id.country_id or tax.country_id
  /home/odoo/odoo19/community/19.0/odoo/orm/fields_relational.py(45)__get__()
-> return super().__get__(records, owner)
  /home/odoo/odoo19/community/19.0/odoo/orm/fields.py(1744)__get__()
-> self.compute_value(recs)
  /home/odoo/odoo19/community/19.0/odoo/orm/fields.py(1915)compute_value()
-> records._compute_field_value(self)
  /home/odoo/odoo19/community/19.0/addons/mail/models/mail_thread.py(484)_compute_field_value()
-> return super()._compute_field_value(field)
  /home/odoo/odoo19/community/19.0/odoo/orm/models.py(4954)_compute_field_value()
-> determine(field.compute, self)
  /home/odoo/odoo19/community/19.0/odoo/orm/fields.py(81)determine()
-> return needle(*args)
> /home/odoo/odoo19/community/19.0/addons/account/models/account_tax.py(296)_compute_tax_group_id()
-> taxes.tax_group_id = self.env['account.tax.group'].search([
```

In the compute of tax_group_id : 
https://github.com/odoo/odoo/blob/19039d174348660007f6d8f23311620f94acef68/addons/account/models/account_tax.py#L284-L300

We need the country_id which won't be passed here, which will setting the tax_group_id to NULL.

Without this changes, first the tax.country_id is computed and then tax_group, so the country_id was passed eventually. 

Till now in this edge case, always account_fiscal_country_id was first assigned to the tax and then the relevant tax_group was fetched, therefore instead of `account_tax.country_id` we could probably fetch the `company_id.account_fiscal_country_id` in `_compute_tax_group_id` as it is the only case(most probably) where this compute is called. That is the one probable fix here. Or we could move back our implementation to the create method. Please let me know your opinion on this one.☺️

**Comment by epictete** [comment]
Thanks for the detailed testing and feedback :+1: 
It's indeed a bit more tricky than it seemed :blush: 
If we add an inverse method to `tax_group_id` in order to set the tax group country to the tax, it solves the current issue, while keeping the solution clean:
```py
    def _inverse_tax_group_id(self):
        for tax in self:
            if tax.tax_group_id.country_id:
                tax.country_id = tax.tax_group_id.country_id
```
Adding logic to the `create()` method is more error prone and should generally be kept as a last resort, if all else fails :nerd_face:

**Comment by kdes-odoo** [comment]
Hey @epictete !!!

Sorry for the delay, I was just trying different solutions for this case, but indeed creating invert is the optimal option here. I've also added some testcases for the same. 

Now regarding the failing testcase here, `test_renting_taxes_included_multicompany()` It just updates the company of tax to check if it will still be applied or not on the SOL. 

Here is the trimmed call stack : 

```python
/home/odoo/odoo19/enterprise/17.0/sale_renting/tests/test_rental.py(555)test_renting_taxes_included_multicompany()
-> self.tax_included.company_id = company2
  /home/odoo/odoo19/community/17.0/odoo/fields.py(1332)__set__()
-> records.write({self.name: write_value})
  /home/odoo/odoo19/community/17.0/addons/account/models/account_tax.py(550)write()
-> return super().write(self._sanitize_vals(vals))
  /home/odoo/odoo19/community/17.0/addons/mail/models/mail_thread.py(325)write()
-> result = super(MailThread, self).write(values)
  /home/odoo/odoo19/community/17.0/odoo/models.py(4476)write()
-> real_recs._validate_fields(vals, inverse_fields)
  /home/odoo/odoo19/community/17.0/odoo/models.py(1468)_validate_fields()
-> check(self)
  /home/odoo/odoo19/community/17.0/addons/account/models/account_tax.py(220)_constrains_name()
-> ('country_id', '=', tax.country_id.id),
  /home/odoo/odoo19/community/17.0/addons/account/models/account_tax.py(242)_compute_country_id()
-> tax.country_id = tax.tax_group_id.country_id or tax.company_id.account_fiscal_country_id or tax.company_id.country_id or tax.country_id
> /home/odoo/odoo19/community/17.0/addons/account/models/account_tax.py(247)_compute_tax_group_id()
-> by_country_company = defaultdict(self.browse)

``` 

Earlier `tax_group_id` was not accessed in `_compute_country_id` so its compute was never called, but now it returns Null which raises error. 

Could you please review the fix and the suggest if I shall adjust this testcase or we make changes in fix ?

Thanks in Advance ☺️
